### PR TITLE
materialize: 0.17.0 -> 0.84.0

### PR DESCRIPTION
**Pull request by happysalada**
## Description of changes

This is still not building yet, but it's very close. I've got some troubles with the npm deps.
@petrosagg in case you have ideas, this is close to ready. (leaving this in draft in the meanwhile).

## Things done

- Built on platform(s)
  - [ ] x86_64-linux
  - [ ] aarch64-linux
  - [ ] x86_64-darwin
  - [ ] aarch64-darwin
- For non-Linux: Is sandboxing enabled in `nix.conf`? (See [Nix manual](https://nixos.org/manual/nix/stable/command-ref/conf-file.html))
  - [ ] `sandbox = relaxed`
  - [ ] `sandbox = true`
- [ ] Tested, as applicable:
  - [NixOS test(s)](https://nixos.org/manual/nixos/unstable/index.html#sec-nixos-tests) (look inside [nixos/tests](https://github.com/NixOS/nixpkgs/blob/master/nixos/tests))
  - and/or [package tests](https://nixos.org/manual/nixpkgs/unstable/#sec-package-tests)
  - or, for functions and "core" functionality, tests in [lib/tests](https://github.com/NixOS/nixpkgs/blob/master/lib/tests) or [pkgs/test](https://github.com/NixOS/nixpkgs/blob/master/pkgs/test)
  - made sure NixOS tests are [linked](https://nixos.org/manual/nixpkgs/unstable/#ssec-nixos-tests-linking) to the relevant packages
- [ ] Tested compilation of all packages that depend on this change using `nix-shell -p nixpkgs-review --run "nixpkgs-review rev HEAD"`. Note: all changes have to be committed, also see [nixpkgs-review usage](https://github.com/Mic92/nixpkgs-review#usage)
- [ ] Tested basic functionality of all binary files (usually in `./result/bin/`)
- [24.05 Release Notes](https://github.com/NixOS/nixpkgs/blob/master/nixos/doc/manual/release-notes/rl-2405.section.md) (or backporting [23.05](https://github.com/NixOS/nixpkgs/blob/master/nixos/doc/manual/release-notes/rl-2305.section.md) and [23.11](https://github.com/NixOS/nixpkgs/blob/master/nixos/doc/manual/release-notes/rl-2311.section.md) Release notes)
  - [ ] (Package updates) Added a release notes entry if the change is major or breaking
  - [ ] (Module updates) Added a release notes entry if the change is significant
  - [ ] (Module addition) Added a release notes entry if adding a new NixOS module
- [ ] Fits [CONTRIBUTING.md](https://github.com/NixOS/nixpkgs/blob/master/CONTRIBUTING.md).

---

Add a :+1: [reaction] to [pull requests you find important].

[reaction]: https://github.blog/2016-03-10-add-reactions-to-pull-requests-issues-and-comments/
[pull requests you find important]: https://github.com/NixOS/nixpkgs/pulls?q=is%3Aopen+sort%3Areactions-%2B1-desc

### CODE REVIEW

File: pkgs/servers/sql/materialize/default.nix [hunk 1]
```diff
@@ -20,10 +25,10 @@ let
...
     };
 
-    static = "./src/materialized/src/http/static";
-    cssVendor = "./src/materialized/src/http/static/css/vendor";
-    jsProdVendor = "./src/materialized/src/http/static/js/vendor";
-    jsDevVendor = "./src/materialized/src/http/static-dev/js/vendor";
+    static = "./src/http/static";
+    cssVendor = "./src/http/static/css/vendor";
+    jsProdVendor = "./src/http/static/js/vendor";
+    jsDevVendor = "./src/http/static-dev/js/vendor";
```
Comment (petrosagg) [inline pkgs/servers/sql/materialize/default.nix:31]: These are the correct paths for the npm deps:

```suggestion
    static = "./src/environmentd/src/http/static";
    cssVendor = "./src/environmentd/src/http/static/css/vendor";
    jsProdVendor = "./src/environmentd/src/http/static/js/vendor";
    jsDevVendor = "./src/environmentd/src/http/static-dev/js/vendor";
```

**Review by petrosagg — COMMENTED** [review]
Thanks for working on this @happysalada! You have forgotten to include the updated `Cargo.lock` in this PR but I reproduced your npm dep issue locally after manually updating the lock file. I put a comment with the correct paths that will get your build going. However I'm still unable to build due to linking issues with RocksDB:

```
error: builder for '/nix/store/2gj7b1ajszr3fng0kmprgvj8zab43rs8-materialize-0.84.0.drv' failed with exit code 101;
       last 10 log lines:
       >           /build/cargo-vendor-dir/rocksdb-0.21.0/src/write_buffer_manager.rs:113: undefined reference to `rocksdb_write_buffer_manager_memory_usage'
       >           /nix/store/xdqs45iclhp9dz8zz9pvn5zivjbhid1a-binutils-2.40/bin/ld: /build/source/target/x86_64-unknown-linux-gnu/debug/deps/librocksdb-48d193e1b7146038.rlib(rocksdb-48d193e1b7146038.rocksdb.223e1bdca97038a7-cgu.06.rcgu.o): in function `<rocksdb::write_buffer_manager::WriteBufferManager>::buffer_size':
       >           /build/cargo-vendor-dir/rocksdb-0.21.0/src/write_buffer_manager.rs:123: undefined reference to `rocksdb_write_buffer_manager_buffer_size'
       >           collect2: error: ld returned 1 exit status
```

**Comment by happysalada** [comment]
@petrosagg I've added the missing cargo.lock file, and made the change, thank you for taking the time.
the rocksdb issue looks like it's more of a problem. The rust rocksdb crate is quite old and only takes care of older rockdb version. We have to wait for an update of that crate, to then make an update in nixpkgs of the rocksdb crate. 
the rust rocksdb crate supports rocksdb 8.1 and currently nixkpkgs has rocksdb 8.3.2.
so we have to wait until everyone can update to 8.10.
I'll leave this in draft for now.
thanks again!

**Comment by petrosagg** [comment]
@happysalada the weird thing is that when developing materialize locally I don't have rocksdb in my system nor do I link it dynamically. So think something is causing `cargo` to not build `librockdb-sys` from source and instead to attempt to link it

**Comment by petrosagg** [comment]
@happysalada I spent some more time on this. Here is the diff that will build the required binaries correctly although I only tried with `doCheck = false;`. You might need to adjust a few tests if you set that to true

```diff
diff --git a/pkgs/servers/sql/materialize/default.nix b/pkgs/servers/sql/materialize/default.nix
index b9c22fec98ea..f4bba8365494 100644
--- a/pkgs/servers/sql/materialize/default.nix
+++ b/pkgs/servers/sql/materialize/default.nix
@@ -13,7 +13,6 @@
 , pkg-config
 , protobuf
 , libclang
-, rocksdb
 , rdkafka
 }:
 
@@ -25,16 +24,20 @@ let
       inherit hash;
     };
 
-    static = "./src/http/static";
-    cssVendor = "./src/http/static/css/vendor";
-    jsProdVendor = "./src/http/static/js/vendor";
-    jsDevVendor = "./src/http/static-dev/js/vendor";
-
     files = with args; [
-      { src = js_prod_file; dst = "${jsProdVendor}/${name}.js"; }
-      { src = js_dev_file;  dst = "${jsDevVendor}/${name}.js"; }
-    ] ++ lib.optional (args ? css_file) { src = css_file; dst = "${cssVendor}/${name}.css"; }
-      ++ lib.optional (args ? extra_file) { src = extra_file.src; dst = "${static}/${extra_file.dst}"; };
+      { src = js_prod_file; dst = "./src/environmentd/src/http/static/js/vendor/${name}.js"; }
+      { src = js_prod_file; dst = "./src/prof-http/src/http/static/js/vendor/${name}.js"; }
+      { src = js_dev_file;  dst = "./src/environmentd/src/http/static-dev/js/vendor/${name}.js"; }
+      { src = js_dev_file;  dst = "./src/prof-http/src/http/static-dev/js/vendor/${name}.js"; }
+    ]
+    ++ lib.optionals (args ? css_file) [
+      { src = css_file; dst = "./src/environmentd/src/http/static/css/vendor/${name}.css"; }
+      { src = css_file; dst = "./src/prof-http/src/http/static/css/vendor/${name}.css"; }
+    ]
+    ++ lib.optionals (args ? extra_file) [
+      { src = extra_file.src; dst = "./src/environmentd/src/http/static/${extra_file.dst}"; }
+      { src = extra_file.src; dst = "./src/prof-http/src/http/static/${extra_file.dst}"; }
+    ];
   in
     lib.concatStringsSep "\n" (lib.forEach files ({src, dst}: ''
       mkdir -p "${dirOf dst}"
@@ -69,8 +72,6 @@ rustPlatform.buildRustPackage rec {
   env.PROTOC = "${protobuf}/bin/protoc";
   env.PROTOC_INCLUDE = "${protobuf}/include";
   env.LIBCLANG_PATH = "${libclang.lib}/lib";
-  env.ROCKSDB_INCLUDE_DIR = "${rocksdb}/include";
-  env.ROCKSDB_LIB_DIR = "${rocksdb}/lib";
   # needed to dynamically link rdkafka
   env.CARGO_FEATURE_DYNAMIC_LINKING=1;
 
@@ -122,8 +123,7 @@ rustPlatform.buildRustPackage rec {
     "--skip test_tls"
   ];
 
-
-  cargoBuildFlags = [ "--bin mz" ];
+  cargoBuildFlags = [ "--bin environmentd --bin clusterd" ];
 
   postInstall = ''
     install --mode=444 -D ./misc/dist/materialized.service $out/etc/systemd/system/materialized.service
```

**Comment by petrosagg** [comment]
> why only build environmentd and clusterd, are those the only two binaries needed to run materialized 

Yeah, these two are the only ones. Specifically you run `environmentd` and it will spawn `clusterd` processes accordingly although there are a bunch of flags need to be passed in. It'd be useful to include a script/service that brings up everything but that can be a separate PR

**Comment by happysalada** [comment]
@petrosagg I think I've got all the requested changes, let me know when you have time to have another look.

**Comment by happysalada** [comment]
Result of `nixpkgs-review pr 282699` run on x86_64-linux [1](https://github.com/Mic92/nixpkgs-review)
<details>
  <summary>1 package built:</summary>
  <ul>
    <li>materialize</li>
  </ul>
</details>

**Comment by happysalada** [comment]
I went ahead and merged, but if you have any other feedback later on, feel free to just post a comment and I'll go for it.

**Comment by petrosagg** [comment]
Looks great, thanks!

**Pull request merged**